homeassistant/components/tts/__init__.py

```diff
@@ -0,0 +1,143 @@
+                          platform)
+            continue
+
+        def generate_speech(tts_service, platform_name, call):
+            """Handle sending tts message service calls."""
+            text = call.data.get(ATTR_TEXT)
+            entity_id = call.data.get(ATTR_ENTITY_ID)
+            allow_cached_file = call.data.get(ATTR_ALLOW_CACHE, True)
+            play_file = call.data.get(ATTR_PLAY, True)
+            language = call.data.get(ATTR_LANGUAGE, config.get(ATTR_LANGUAGE,
+                                                               "en-us"))
+            rate = call.data.get(ATTR_RATE, config.get(ATTR_RATE, 0))
+            codec = call.data.get(ATTR_CODEC, config.get(ATTR_CODEC, "MP3"))
+            audio_format = call.data.get(ATTR_FORMAT,
+                                         config.get(ATTR_FORMAT,
+                                                    "44khz_16bit_stereo"))
+
+            hashed_text = hashlib.sha1(text.encode('utf-8')).hexdigest()
+
+            full_filename = "{}-{}-{}.{}".format(hashed_text, language,
+                                                 platform_name,
+                                                 codec.lower())
+
+            file_path = os.path.join(tts_dir, full_filename)
+
+            if text is None:
+                _LOGGER.error(
+                    'Received call to %s without attribute %s',
+                    call.service, ATTR_TEXT)
+                return
+
+            if (os.path.isfile(file_path)) is False or \
+               (allow_cached_file is False):
+                try:
+                    tts_service.speak(file_path, text, language, rate, codec,
+                                      audio_format)
+                # pylint: disable=bare-except
+                except:
+                    _LOGGER.error('Error trying to speak using %s',
+                                  platform_name)
+                    return
+
+            if play_file is True or entity_id is None:
+                content_url = "{}/api/tts/{}".format(hass.config.api.base_url,
+                                                     full_filename)
+                hass.services.call(core.DOMAIN, SERVICE_PLAY_MEDIA, {
+                    ATTR_ENTITY_ID: entity_id,
+                    ATTR_MEDIA_CONTENT_ID: content_url,
+                    ATTR_MEDIA_CONTENT_TYPE: MEDIA_TYPE_MUSIC
+                })
+
+        service_tts = p_config.get(CONF_NAME, platform)
+        service_call_handler = partial(generate_speech, tts_engine,
+                                       service_tts)
+        hass.services.register(DOMAIN, service_tts, service_call_handler,
+                               descriptions.get(SERVICE_TTS))
+        success = True
+
+    hass.http.register_path(
+        'GET', re.compile(r'/api/tts/(?P<tts_filename>.*)'),
+        _handle_get_tts_file)
+
+    return success
+
+
+def _handle_get_tts_file(handler, path_match, data):
```

Comment:
Idea for future improvement: if the tts_filename starts with a specific prefix have the file be deleted after being served to the media player.